gensim/corpora/_mmreader.pyx

```diff
@@ -0,0 +1,175 @@
+from __future__ import with_statement
+
+from gensim import utils
+
+from six import string_types
+from six.moves import xrange
+import logging
+
+cimport cython
+from libc.stdio cimport sscanf
+
+
+logger = logging.getLogger(__name__)
+
+
+cdef class MmReader(object):
```

Comment:
This is almost same as a pure-python version (`sscanf` different), maybe something else?

Reply:
No, nothing else really.  It's a very trivial change.  Just a change to use sscanf instead of split() and int() and a few type definitions (eg, docid, termid, and previd in __iter__).

No doubt converting the actual file access to c would further improve performance, but that will require dealing with all the different file types that smart_open accepts.  Also, I don't think this is the main hotspot in __iter__.

For example, here's a line profiler output of the current (develop branch) code when reading a raw text file:

<pre>
Line #  % Time  Line Contents
==============================
   907     0.0          with utils.file_or_filename(self.input) as lines:
   908     0.0              self.skip_headers(lines)

   911     8.6              for line in lines:
   912    40.4                  docid, termid, val = utils.to_unicode(line).split()  # needed for python3
   913     7.0                  if not self.transposed:
   914                              termid, docid = docid, termid
   915                          # -1 because matrix market indexes are 1-based => convert to 0-based
   916    22.5                  docid, termid, val = int(docid) - 1, int(termid) - 1, float(val)
   917     6.7                  assert previd <= docid, "matrix columns must come in ascending order"
   918     6.4                  if docid != previd:

   931
   932     8.4                  document.append((termid, val,))  # add another field to the current document
   933
</pre>
The majority of time is spent splitting the line and converting the resulting strings to ints (lines 912 and 916).  sscanf handles both of these lines.  Another 20% or so is taken up by the testing self.transposed and comparing docid and previd (lines 913, 917, and 918).   Statically typing these variables in cython helps quite a bt too.

For a compressed text file, line 911 (for line in lines) takes up 33% of the time, but most of the time is still taken by line.split() and int().

Reply:
Here's the line profile output when reading a compressed file

<pre>
Line #  % Time  Line Contents
==============================
   907     0.0          with utils.file_or_filename(self.input) as lines:
   908     0.0              self.skip_headers(lines)

   911    31.5              for line in lines:
   912    30.6                  docid, termid, val = utils.to_unicode(line).split()  # needed for python3
   913     5.1                  if not self.transposed:
   914                              termid, docid = docid, termid
   915                          # -1 because matrix market indexes are 1-based => convert to 0-based
   916    17.0                  docid, termid, val = int(docid) - 1, int(termid) - 1, float(val)
   917     4.9                  assert previd <= docid, "matrix columns must come in ascending order"
   918     4.6                  if docid != previd:

   932     6.3                  document.append((termid, val,))  # add another field to the current document
</pre>